flyteadmin/dataproxy/service.go

```diff
@@ -117,23 +128,33 @@ func (s Service) CreateUploadLocation(ctx context.Context, req *service.CreateUp
 	storagePath, err := createStorageLocation(ctx, s.dataStore, s.cfg.Upload,
 		req.Project, req.Domain, prefix, req.Filename)
 	if err != nil {
+		logger.Errorf(ctx, "failed to create shardedStorageLocation. Error %v", err)
 		return nil, errors.NewFlyteAdminErrorf(codes.Internal, "failed to create shardedStorageLocation, Error: %v", err)
 	}
 
+	metadata := make(map[string]string)
+
+	if req.AddMetadata {
+		metadata[storage.FlyteContentMD5] = base64digestMD5
+	}
```

Comment:
1. I think `dataStore.CreateSignedURL()` should be the one that adds this metadata... ideally you want to isolate assumptions here. If the storage package assumes this tag is set and its value is the contentMD5, then it should be the one that adds it as well.

Reply:
2. Is the reason for having `req.AddMetadata` to allow backward compatibility? Added more notes in the overall review for how to handle backward compatibility. Discussed this with Eduardo earlier.